File: backend/services/session.js
const { v4: uuidv4 } = require('uuid');
const { STATES } = require('../utils/stateManager');

// ─── In-memory store — intentionally no database ──────────────────────────────
// All application data lives here until emailed to the owner.
// Sessions auto-expire after 30 minutes of inactivity.
const sessions = new Map();

const SESSION_TTL = 30 * 60 * 1000; // 30 minutes

// ─── Create ───────────────────────────────────────────────────────────────────
const createSession = () => {
  const sessionId = uuidv4();

  const session = {
    sessionId,
    state: STATES.JOB_SELECTION, // First real collection state (greeting is handled by /start)
    data: {
      jobRole:        null,
      location:       null,
      shift:          null,
      name:           null,
      email:          null,
      phone:          null,
      dob:            null,
      startDate:      null,
      resumeBuffer:   null,  // Buffer — kept in memory, attached to email
      resumeFileName: null,
      linkedIn:       null,
    },
    history:    [],   // Last N conversation turns fed to Claude for context
    createdAt:  Date.now(),
    lastActive: Date.now(),
  };

  sessions.set(sessionId, session);
  return session;
};

// ─── Read ─────────────────────────────────────────────────────────────────────
const getSession = (sessionId) => {
  if (!sessionId) return null;
  const session = sessions.get(sessionId);
  if (!session) return null;

  // Evict expired sessions on access
  if (Date.now() - session.lastActive > SESSION_TTL) {
    sessions.delete(sessionId);
    return null;
  }

  session.lastActive = Date.now();
  return session;
};

// ─── Update ───────────────────────────────────────────────────────────────────
const updateSession = (sessionId, { state, data } = {}) => {
  const session = sessions.get(sessionId);
  if (!session) return null;

  if (state) session.state = state;
  if (data) Object.assign(session.data, data);
  session.lastActive = Date.now();

  sessions.set(sessionId, session);
  return session;
};

// ─── Conversation history ─────────────────────────────────────────────────────
const addToHistory = (sessionId, role, content) => {
  const session = sessions.get(sessionId);
  if (!session) return;

  session.history.push({ role, content });

  // Keep a rolling window — Claude doesn't need the full transcript
  if (session.history.length > 20) {
    session.history = session.history.slice(-20);
  }

  session.lastActive = Date.now();
  sessions.set(sessionId, session);
};

// ─── Cleanup job — runs every 10 minutes ─────────────────────────────────────
setInterval(() => {
  const now = Date.now();
  let evicted = 0;
  for (const [id, session] of sessions.entries()) {
    if (now - session.lastActive > SESSION_TTL) {
      sessions.delete(id);
      evicted++;
    }
  }
  if (evicted > 0) console.log(`[Session] Evicted ${evicted} expired session(s). Active: ${sessions.size}`);
}, 10 * 60 * 1000);

module.exports = { createSession, getSession, updateSession, addToHistory };

File: backend_py/services/session.py
import uuid
import time
import threading
from typing import Optional, Dict

# ─── Config ───────────────────────────────────────────────────────────────────
SESSION_TTL = 30 * 60  # 30 minutes in seconds

# ─── In-memory store ──────────────────────────────────────────────────────────
_sessions: Dict[str, dict] = {}
_lock = threading.Lock()


# ─── Session shape ────────────────────────────────────────────────────────────
def _new_data() -> dict:
    return {
        "jobRole":        None,
        "location":       None,
        "shift":          None,
        "name":           None,
        "email":          None,
        "phone":          None,
        "dob":            None,
        "startDate":      None,
        "resumeBuffer":   None,
        "resumeFileName": None,
        "linkedIn":       None,
    }


# ─── CRUD ─────────────────────────────────────────────────────────────────────
def create_session() -> dict:
    session_id = str(uuid.uuid4())
    session = {
        "sessionId":  session_id,
        "state":      "JOB_SELECTION",
        "data":       _new_data(),
        "history":    [],
        "createdAt":  time.time(),
        "lastActive": time.time(),
    }
    with _lock:
        _sessions[session_id] = session
    return session


def get_session(session_id: str) -> Optional[dict]:
    with _lock:
        session = _sessions.get(session_id)
        if not session:
            return None
        # evict if expired
        if time.time() - session["lastActive"] > SESSION_TTL:
            del _sessions[session_id]
            return None
        session["lastActive"] = time.time()
        return session


def update_session(session_id: str, updates: dict) -> Optional[dict]:
    with _lock:
        session = _sessions.get(session_id)
        if not session:
            return None
        if "state" in updates:
            session["state"] = updates["state"]
        if "data" in updates:
            for key, val in updates["data"].items():
                session["data"][key] = val
        session["lastActive"] = time.time()
        return session


def add_to_history(session_id: str, role: str, content: str):
    with _lock:
        session = _sessions.get(session_id)
        if session:
            session["history"].append({"role": role, "content": content})
            session["lastActive"] = time.time()


# ─── Cleanup ──────────────────────────────────────────────────────────────────
def _cleanup():
    with _lock:
        now = time.time()
        expired = [sid for sid, s in _sessions.items() if now - s["lastActive"] > SESSION_TTL]
        for sid in expired:
            del _sessions[sid]
    if expired:
        print(f"[Session] Cleaned up {len(expired)} expired session(s)")


def start_cleanup_scheduler():
    """Runs session cleanup every 10 minutes in a background daemon thread."""
    def run():
        while True:
            time.sleep(10 * 60)
            _cleanup()
    thread = threading.Thread(target=run, daemon=True)
    thread.start()
